[GEN][ZH] Implement REPLAY NOT FOUND and MAP NOT FOUND dialogs when attempting to load a Replay that was deleted or whose map is missing

xezon opened the pull request:
**Merge with Rebase**

* Merge after #993
* Fixes #126
* Follow up for #984
* Split off of #398

This change implements a new MAP NOT FOUND dialog when the user attempts to load a Replay whose map is missing in the user data (or game install). Originally the game would crash when trying to start a new game with a missing map.

Additionally it implements a new REPLAY NOT FOUND dialog when the user attempts to load a Replay that was deleted after the Replay Menu was opened.

This change has 2 commits. The first commit is a refactor to help implement the fix efficiently. The second commit contains new code for GameText to allow looking up localization texts with substitute texts when those localizations are not found (which they currently are not because we do not touch game data). The substitution code can be compiled out cleanly when so desired (via GameDefines.h).

## TODO

- [x] Replicate in Generals

## Tip

Ignore whitespace in diff viewer to ease reviewing this change.

## Dialog Image

![shot_20250601_164828_1](https://github.com/user-attachments/assets/b1d8e9cc-c464-4468-88a2-7da5b3d6a5ac)

helmutbuhler commented on Core/GameEngine/Include/Common/GameDefines.h:
```diff
@@ -28,3 +28,7 @@
 #endif
+
+#ifndef ENABLE_GAMETEXT_SUBSTITUTES
+#define ENABLE_GAMETEXT_SUBSTITUTES (1) // The code can provide substitute texts when labels and strings are missing in the STR or CSF translation file
```
Do we really need this? Why would we need the ability to compile out the substitutions?

roossienb replied:
I agree with @helmutbuhler here. I don't think it should be possible to disable the substitution. 
I find using keys (such as `GUI:LabelName`) to be old fashioned. 

In C# for example, the actual English text is used as a key to look up other translations. If the translation is missing, it falls back to the English text. This ensures the user never gets to see a weird key code.

xezon replied:
> Do we really need this?

Yes. Disabling this gives nice fat MISSING:"GUI:Label" texts in the GUI which is useful when reviewing it.

helmutbuhler replied:
Hm, I still don't understand. Why would you want to review it with substitutions disabled? Even if there is a reason, you can just go to `GameTextManager::fetchOrSubstitute` and comment out:
```cpp
	if (!exists)
		str = substituteText;
```

xezon replied:
Commenting out the code will still compile in the substitution strings in the executable. This means it is not a clean strip out. Also finding the right place to comment out is not very user friendly. Ideally it is a compile option and/or command line argument.

GameDefines.h is perfectly fine for such things.

helmutbuhler commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp:
```diff
@@ -102,6 +102,44 @@ UnicodeString GetReplayFilenameFromListbox(GameWindow *listbox, Int index)
+	ReplayMapInfo() : md(NULL) {}
+
+	bool headerAndInfoOk() const { return info.getMapSize() != 0; }
+	bool mdOk() const { return md != NULL; }
```
`mapExists` instead of `mdOk`?

roossienb replied:
See my suggestions on line 114. I would prefer `hasMap` or `hasMetadata` or `isValidMap` rather than `Ok`
Also for the line above: `hasHeader` or `isValidHeader`

xezon replied:
Function removed.

helmutbuhler reviewed (approved):
Looks good. I would have preferred out parameters instead of the ReplayMapInfo struct, which is a bit weird because it has methods as a struct, doesn't have m_ prefix for the members and makes callsite code more complicated.

helmutbuhler commented:
Btw, can you also 'polish' PR #760? It's also split off of #398, and I don't know how to get your approval there.

roossienb commented on GeneralsMD/Code/GameEngine/Include/GameClient/GameText.h:
```diff
@@ -92,5 +94,14 @@ extern GameTextInterface* CreateGameTextInterface( void );
+#else
+#define FETCH_OR_SUBSTITUTE(labelA, substituteTextW) fetch(labelA)
+#endif
+
```
Not needed when substitute cannot be compiled out.

roossienb commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp:
```diff
@@ -102,6 +102,44 @@ UnicodeString GetReplayFilenameFromListbox(GameWindow *listbox, Int index)
+	bool headerAndInfoOk() const { return info.getMapSize() != 0; }
+	bool mdOk() const { return md != NULL; }
+
+	const MapMetaData* md;
```
`md` is quite a cryptic name for a variable. I suggest using `metaData` or `mapMetaData` or maybe even `map`

roossienb commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp:
```diff
@@ -102,6 +102,44 @@ UnicodeString GetReplayFilenameFromListbox(GameWindow *listbox, Int index)
+
+static ReplayMapInfo readReplayMapInfo(const AsciiString& filename)
+{
+	ReplayMapInfo rmi;
```
Cryptic name. I suggest `replayMap`

roossienb commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp:
```diff
@@ -102,6 +102,44 @@ UnicodeString GetReplayFilenameFromListbox(GameWindow *listbox, Int index)
+{
+	ReplayMapInfo() : md(NULL) {}
+
+	bool headerAndInfoOk() const { return info.getMapSize() != 0; }
```
Is it possible that `info` is NULL? In that case, nullcheck must be performed.

roossienb commented on GeneralsMD/Code/GameEngine/Source/GameClient/GameText.cpp:
```diff
@@ -1341,6 +1343,19 @@ UnicodeString GameTextManager::fetch( AsciiString label, Bool *exists )
+	UnicodeString str = fetch(label, &exists);
+	if (!exists)
+		str = substituteText;
+	return str;
```
Add an empty line after the unbraced if statement.

roossienb commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp:
```diff
@@ -449,7 +481,18 @@ static void loadReplay(UnicodeString filename)
 	asciiFilename.translate(filename);
 
-	if(!TheRecorder->replayMatchesGameVersion(asciiFilename))
+	ReplayMapInfo rmi = readReplayMapInfo(asciiFilename);
```
Variable naming

roossienb commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp:
```diff
@@ -145,124 +183,118 @@ void PopulateReplayFileListbox(GameWindow *listbox)
-				// version (no-op)
+			// map
+			UnicodeString mapStr;
+			const MapMetaData *md = TheMapCache->findMap(rmi.info.getMap());
```
Variable naming

xezon replied:
Fixed.

roossienb reviewed (changes requested):
With the subsitute in place, it would be great to have a follow up PR with:
- add text for when the replay is paused/unpaused
- Use substitute for Fast-forward message in Generals

xezon commented:
> With the subsitute in place, it would be great to have a follow up PR with:
> * add text for when the replay is paused/unpaused
> * Use substitute for Fast-forward message in Generals

Yes. Good idea.

xezon commented:
> I would have preferred out parameters instead of the ReplayMapInfo struct, which is a bit weird because it has methods as a struct, doesn't have m_ prefix for the members and makes callsite code more complicated.

Ok changed to function.

I also moved a bit more of the code into separate functions that could be reused elsewhere.

I also noticed that it is possible to attempt loading a Replay file that was deleted after the Replay Menu was opened. It then triggered the OLDER REPLAY VERSION dialog with the OK button. Pressing OK would get the game stuck loading nothing. So I added another new REPLAY NOT FOUND message to prevent that.

![shot_20250602_191632_1](https://github.com/user-attachments/assets/209f1874-df83-49be-b96c-0269c16b8639)

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/ReplayMenu.cpp:
```diff
@@ -102,6 +102,75 @@ UnicodeString GetReplayFilenameFromListbox(GameWindow *listbox, Int index)
+	{
+		// TheSuperHackers @bugfix helmutbuhler 08/03/2025 Just use the filename.
+		// Displaying a long map path string would break the map list gui.
+		const char* filename = info.getMap().reverseFind('\\');
```
Note to myself: I suspect this will still break the Replay list if the file name is very long then. So we perhaps need another change that caps the length of the Replay name here. Unrelated to this Pull Request.

xezon commented:
Replicated to Generals without conflicts. Commit title improved.
